src/unix/battery.c

```diff
@@ -0,0 +1,252 @@
+
+  if (pCFRelease == NULL ||
+      pCFStringCreateWithCString == NULL ||
+      pCFArrayGetCount == NULL ||
+      pCFArrayGetValueAtIndex == NULL ||
+      pCFNumberGetValue == NULL ||
+      pCFDictionaryGetValue == NULL ||
+      pIOPSCopyPowerSourcesInfo == NULL ||
+      pIOPSCopyPowerSourcesList == NULL ||
+      pIOPSGetPowerSourceDescription == NULL) {
+    goto out;
+  }
+
+#define S(s) pCFStringCreateWithCString(NULL, (s), kCFStringEncodingUTF8)
+
+  blob = pIOPSCopyPowerSourcesInfo();
+  if (blob == NULL)
+    goto out;
+
+  sources = pIOPSCopyPowerSourcesList(blob);
+  if (sources == NULL)
+    goto out;
+
+  // If there are no sources, we're on a desktop or server, nothin' to do
+  if (pCFArrayGetCount(sources) == 0) {
+    err = UV_EPERM;
+    goto out;
+  }
+
+  powerSource =
+    pIOPSGetPowerSourceDescription(blob, pCFArrayGetValueAtIndex(sources, 0));
+  if (powerSource == NULL)
+    goto out;
+
+  err = 0;
+
+  // kIOPSCurrentCapacityKey
+  currentCapacity = pCFDictionaryGetValue(powerSource, S("Current Capacity"));
+  if (currentCapacity != NULL) {
+    pCFNumberGetValue(currentCapacity, pkCFNumberIntType, &holder);
+    info->level = holder;
+  }
+
+
+  // Retrieve is_charging status
+  // kIOPSIsChargingKey
+  isChargingValue = pCFDictionaryGetValue(powerSource, S("Is Charging"));
+  if (isChargingValue != NULL) {
+    is_charging = pCFBooleanGetValue(isChargingValue);
+    info->is_charging = is_charging;
+  }
+
+  // kIOPSTimeToFullChargeKey and kIOPSTimeToEmptyKey
+  timeRemainingValue =
+    pCFDictionaryGetValue(powerSource,
+        is_charging ? S("Time to Full Charge") : S("Time to Empty"));
+  if (timeRemainingValue != NULL) {
+    pCFNumberGetValue(timeRemainingValue, pkCFNumberIntType, &holder);
+  }
+
+  // A value of -1 indicates "Still Calculating the Time"
+  if (holder == -1)
+    goto out;
+
+  holder = holder * 60; // Default unit is minutes, convert to seconds
+  if (is_charging == 1) {
+    info->charge_time_in_secs = holder;
+    info->discharge_time_in_secs = 0;
+  } else {
+    info->discharge_time_in_secs = holder;
+    info->charge_time_in_secs = 0;
+  }
+out:
+  if (iokit_framekwork_handle != NULL) {
+    dlclose(iokit_framekwork_handle);
+  }
+
+  if (core_foundation_handle != NULL) {
+    dlclose(core_foundation_handle);
+  }
+
+  if (blob != NULL) {
+    pCFRelease(blob);
+  }
+
+  if (sources != NULL) {
+    pCFRelease(sources);
+  }
+#elif defined(__LINUX__)
+  char buf[1024];
+  FILE* fp;
+
+  err = UV_EPERM;
+  fp = uv__open_file("/sys/class/power_supply/BAT0");
+  if (fp == NULL) {
+    goto out;
+  }
+
+  fp = uv__open_file("/sys/class/power_supply/BAT0/status");
+  if (fp == NULL) {
+    goto out;
+  }
+
+  if (fgets(buf, sizeof(buf), fp) == NULL) {
+    goto out;
+  }
+
+  fclose(fp);
+
+  if (strstr(buf, "Charging") != NULL) {
+    info->is_charging = 1;
+    fp = uv__open_file("/sys/class/power_supply/BAT0/time_to_full_now");
```

Comment:
I'm fairly sure time_to_full_now and time_to_empty_now don't exist.

I'm not sure how you'd estimate time-to-full, I don't think the kernel exposes that directly. You're probably supposed to piece it together from charge_full and charge_now, or something like that.

Reply:
I think it does https://github.com/torvalds/linux/blob/master/drivers/power/supply/power_supply_sysfs.c#L209

I validated that with some coworker's machine.

Reply:
Ah, okay. Looks like there are only two or three drivers that implement it. The files don't exist on my system.